### PR TITLE
db: add sqlite "source of truth" schema

### PR DESCRIPTION
**Pull request by kradalby**
This PR is kind of like a "stretch goal". At the moment it does not pass because our migrations does not actually bring various databases into a consistent state. 

This adds a `schema.sql` which should be the source of truth representation of our schema after all migration is applied (for sqlite). This is to ensure you can look one place to figure out how the current schema is defined and not have to do the patchwork of migrations in your head to get an overview.

I suspect we might have strange behaviour and bugs showing up because there exists schemas out there with different constraint than we think they have. In addition, we might have constraint that does not behave the way we expect, simply because we do not have the complete overview of all constraints. #2475 might be a result of this.

  

Thanks to @nblock for discovering this.

### CODE REVIEW

File: hscontrol/db/schema.sql [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  ipv6 text,
+  hostname text,
+  given_name varchar(63),
+  user_id integer,
+  register_method text,
+  forced_tags text,
+  auth_key_id integer,
+  last_seen datetime,
+  expiry datetime,
+
```
Comment (nblock) [inline hscontrol/db/schema.sql:93]: The `nodes` table misses the `approved_routes` column.

File: hscontrol/db/db.go [hunk 2]
```diff
@@ -718,13 +723,366 @@ AND auth_key_id NOT IN (
...
+								ipv4 text,
+								ipv6 text,
+								hostname text,
+								given_name varchar(63),
+								user_id integer,
+								register_method text,
+								forced_tags text,
+								auth_key_id integer,
+								last_seen datetime,
+								expiry datetime,
```
Comment (nblock) [inline hscontrol/db/db.go]: `approved_routes` is missing here.

**Comment by kradalby** [comment]
I'm marking this as ready, but I would like to gather some more "broken" schemas, so will post in discord. also posted on #2475 and #2597.

If you have a broken database, I would love to get a dump of it. The most valuable would be:
- The older (or multiple steps along the way) the better (an old backup for example)
- The more data the better
- Scrub PII, auth keys and API, but don't remove things that don't matter (you can leave public keys for node, disco machine)
- A description of the type of error you got when it was bad, and possibly how you fixed it.

A dump could be taken with:
```
sqlite3 <DATABASE PATH> ".dump" > database.sql
```

If you have one of these, you can email me an SQL dump, my email is in my Github profile.

File: hscontrol/db/node_test.go [hunk 3]
```diff
@@ -118,7 +118,7 @@ func (s *Suite) TestListPeers(c *check.C) {
 	_, err = db.GetNodeByID(0)
 	c.Assert(err, check.NotNil)
 
-	for index := 0; index <= 10; index++ {
+	for index := range 11 {
```
Comment (ghost) [inline hscontrol/db/node_test.go:121]: In hscontrol/db/node_test.go, you're using `for index := range 11` which is valid syntax in Go 1.22+. Make sure your CI environments and developers are using Go 1.22 or newer to avoid compile errors.

File: hscontrol/db/sqliteconfig/config.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	}
+
+	return nil
+}
+
+// ToURL builds a properly encoded SQLite connection string using _pragma parameters
+// compatible with modernc.org/sqlite driver.
+func (c *Config) ToURL() (string, error) {
+	if err := c.Validate(); err != nil {
+		return "", fmt.Errorf("invalid config: %w", err)
```
Comment (ghost) [inline hscontrol/db/sqliteconfig/config.go:302]: In hscontrol/db/sqliteconfig/config.go, consider using a parameter struct for ToURL() to make future extensions easier, as you already have several configuration parameters that could grow over time.

File: hscontrol/db/schema.sql [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  expiry datetime,
+  approved_routes text,
+
+  created_at datetime,
+  updated_at datetime,
+  deleted_at datetime,
+
+  CONSTRAINT fk_nodes_user FOREIGN KEY(user_id) REFERENCES users(id) ON DELETE CASCADE,
+  CONSTRAINT fk_nodes_auth_key FOREIGN KEY(auth_key_id) REFERENCES pre_auth_keys(id)
+);
```
Comment (ghost) [inline hscontrol/db/schema.sql]: In hscontrol/db/schema.sql, consider adding indexes on frequently queried columns in the nodes table (like machine_key, node_key, and hostname) to improve query performance.

File: hscontrol/db/schema.sql [hunk 6]
```diff
@@ -0,0 +1,110 @@
+-- This  file is the representation of the SQLite schema of Headscale.
```
Comment (ghost) [inline hscontrol/db/schema.sql]: In hscontrol/db/schema.sql, the comment on line 1 has a double space after 'This' which is inconsistent with the rest of the comments. Consider fixing for consistency.

File: hscontrol/db/sqliteconfig/config.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+// Errors returned by config validation.
+var (
+	ErrPathEmpty           = errors.New("path cannot be empty")
+	ErrBusyTimeoutNegative = errors.New("busy_timeout must be >= 0")
+	ErrInvalidJournalMode  = errors.New("invalid journal_mode")
+	ErrInvalidAutoVacuum   = errors.New("invalid auto_vacuum")
+	ErrWALAutocheckpoint   = errors.New("wal_autocheckpoint must be >= -1")
+	ErrInvalidSynchronous  = errors.New("invalid synchronous")
+)
```
Comment (ghost) [inline hscontrol/db/sqliteconfig/config.go:19]: In hscontrol/db/sqliteconfig/config.go, the error variables are defined but some lack unit tests. Consider adding tests that validate error handling, especially for the validation logic.

File: hscontrol/db/node_test.go [hunk 8]
```diff
@@ -888,32 +892,32 @@ func TestListNodes(t *testing.T) {
...
 	assert.Equal(t, "test2", nodes[0].Hostname)
 
 	// Several matched IDs
 	nodes, err = db.ListNodes(types.NodeIDs{1, 2, 3}...)
 	require.NoError(t, err)
-	assert.Equal(t, len(nodes), 2)
+	assert.Len(t, nodes, 2)
 	assert.Equal(t, "test1", nodes[0].Hostname)
 	assert.Equal(t, "test2", nodes[1].Hostname)
 }
```
Comment (ghost) [inline hscontrol/db/node_test.go:923]: In hscontrol/db/node_test.go, consider adding test cases that verify handling of empty slices and nil values for the ListNodes function, as these edge cases might occur in production.

File: hscontrol/db/schema.sql [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+-- It is the "source of truth" and is used to validate any migrations
+-- that are run against the database to ensure it ends in the expected state.
+
+CREATE TABLE migrations(id text,PRIMARY KEY(id));
```
Comment (ghost) [inline hscontrol/db/schema.sql:5]: In hscontrol/db/schema.sql, the PRIMARY KEY definition for migrations table uses a different style than other tables. Consider standardizing to either inline or separate constraint declaration style for consistency.

File: hscontrol/db/sqliteconfig/config.go [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+		return true
+	default:
+		return false
+	}
+}
+
+// String returns the string representation.
+func (s Synchronous) String() string {
+	return string(s)
+}
```
Comment (ghost) [inline hscontrol/db/sqliteconfig/config.go]: In hscontrol/db/sqliteconfig/config.go, there are no comments explaining the implications of different journal modes and synchronous settings. Consider adding documentation that clarifies performance/durability tradeoffs for different configurations.

File: hscontrol/db/sqliteconfig/config.go [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+//   - Automatic space management (INCREMENTAL auto-vacuum)
+//   - Data integrity (foreign key constraints enabled)
+//   - Reasonable timeout for busy database scenarios (10s)
+func Default(path string) *Config {
+	return &Config{
+		Path:              path,
+		BusyTimeout:       DefaultBusyTimeout,    // 10s timeout for concurrent access
+		JournalMode:       JournalModeWAL,        // Best concurrency + crash recovery
+		AutoVacuum:        AutoVacuumIncremental, // Gradual space reclamation
+		WALAutocheckpoint: 0,                     // Disabled for better control
```
Comment (nblock) [inline hscontrol/db/sqliteconfig/config.go]: Should this be `1000` to match our configuration in `config-example.yaml` and code?

File: hscontrol/db/db.go [hunk 12]
```diff
@@ -718,13 +733,397 @@ AND auth_key_id NOT IN (
...
+		// or else it blocks...
+		sqlConn.SetMaxIdleConns(maxIdleConns)
+		sqlConn.SetMaxOpenConns(maxOpenConns)
+		defer sqlConn.SetMaxIdleConns(1)
+		defer sqlConn.SetMaxOpenConns(1)
+
+		ctx, cancel := context.WithTimeout(context.Background(), contextTimeoutSecs*time.Second)
+		defer cancel()
+
+		if err := squibble.Validate(ctx, sqlConn, dbSchema); err != nil {
```
Comment (nblock) [inline hscontrol/db/db.go:964]: I deliberately modified the schema and it reported the difference. The error is reported as single line which makes it hard to read: `FTL cmd/headscale/cli/serve.go:24 > Error initializing error="creating new headscale: init state: init database: validating schema: invalid schema (-got, +want):\n\n>> Modify index \"idx_policies_deleted_at\"\n@@ -1 +1 @@\n-CREATE INDEX idx_policies_deleted_at ON \"policies2\"(deleted_at)\n+CREATE INDEX idx_policies_deleted_at ON policies(deleted_at)\n\n>> Remove table \"policies2\"\n\n>> Add table \"policies\"\n+ CREATE TABLE policies(\n+   id integer PRIMARY KEY AUTOINCREMENT,\n+   data text,\n+ \n+   created_at datetime,\n+   updated_at datetime,\n+   deleted_at datetime\n+ )\n"`

Is it possible to print the error on multiple lines?

File: hscontrol/db/db.go [hunk 13]
```diff
@@ -718,13 +733,397 @@ AND auth_key_id NOT IN (
...
+						}
+
+						for nodeID, routes := range tableData.RoutesData {
+							tsaddr.SortPrefixes(routes)
+							tableData.RoutesData[nodeID] = slices.Compact(routes)
+						}
+						log.Info().Int("routes", len(routes)).Int("nodes_with_routes", len(tableData.RoutesData)).Msg("Collected routes data")
+					}
+
+					// Drop all existing tables to avoid foreign key constraint issues
```
Comment (nblock) [inline hscontrol/db/db.go]: This seems dangerous to me. If any of the restore steps below fail, we'd leave the user with an empty/half-filled database.

Have you considered the alternative:

- Rename existing tables to `$TABLE_old`
- Drop all indices if exists
- Restore data
- Drop old tables

Data could be directly inserted from old to the new table (in case no changes are required) and the data is still there in the renamed tables in case anything fails during restore. 

The migration itself could probably not run a second time because old tables are renamed already but at least the data is still there in case the user has no backup.

File: hscontrol/db/db.go [hunk 14]
```diff
@@ -718,13 +733,397 @@ AND auth_key_id NOT IN (
...
+						`CREATE TABLE api_keys(
+  id integer PRIMARY KEY AUTOINCREMENT,
+  prefix text,
+  hash blob,
+  expiration datetime,
+  last_seen datetime,
+
+  created_at datetime
+)`,
+						`CREATE TABLE "nodes"(
```
Comment (nblock) [inline hscontrol/db/db.go]: `CREATE TABLE nodes(`

maybe without quotes for consistency with the other `CREATE TABLE` statements.

File: hscontrol/db/schema.sql [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+  prefix text,
+  hash blob,
+  expiration datetime,
+  last_seen datetime,
+
+  created_at datetime
+);
+CREATE UNIQUE INDEX idx_api_keys_prefix ON api_keys(prefix);
+
+CREATE TABLE "nodes"(
```
Comment (nblock) [inline hscontrol/db/schema.sql]: `CREATE TABLE nodes(`

maybe without quotes for consistency with the other `CREATE TABLE` statements.

**Review by nblock — COMMENTED** [review]
Tested this PR with several databases and they got migrated properly.

File: hscontrol/db/db.go (hunk 13 above)
Comment (kradalby) [reply hscontrol/db/db.go]: Smarter, PTAL

File: hscontrol/db/db.go (hunk 12 above)
Comment (kradalby) [reply hscontrol/db/db.go:964]: Should not be printed more clear in human mode, logging library was not very keen on respecting newlines.

File: CHANGELOG.md [hunk 16]
```diff
@@ -2,13 +2,56 @@
...
+
+Please read the [PR description](https://github.com/juanfont/headscale/pull/2617)
+for more technical details about the issues and solutions.
+
+**SQLite Database Backup Example:**
+```bash
+# Stop headscale
+systemctl stop headscale
+
+# Create backup (includes WAL files if using WAL mode)
```
Comment (nblock) [inline CHANGELOG.md]:
```suggestion
# Backup sqlite database
```

File: CHANGELOG.md [hunk 17]
```diff
@@ -2,13 +2,56 @@
...
+cp /var/lib/headscale/db.sqlite /var/lib/headscale/db.sqlite.backup
+
+# and any WAL related files if they exist
```
Comment (nblock) [inline CHANGELOG.md]:
```suggestion
# Backup sqlite WAL/SHM files (if they exist)

File: hscontrol/db/db.go [hunk 18]
```diff
@@ -718,13 +733,239 @@ AND auth_key_id NOT IN (
...
+					}
+
+					log.Info().Msg("Schema recreation completed successfully")
+					return nil
+				},
+				Rollback: func(db *gorm.DB) error { return nil },
+			},
+			// From this point, the following rules must be followed:
+			// - NEVER use gorm.AutoMigrate, write the exact migration steps needed
+			// 	 - AutoMigrate depends on the struct staying exactly the same, which it wont over time.
```
Comment (nblock) [inline hscontrol/db/db.go]:
```suggestion
			// - AutoMigrate depends on the struct staying exactly the same, which it wont over time.
```

File: hscontrol/db/db.go [hunk 19]
```diff
@@ -718,13 +733,239 @@ AND auth_key_id NOT IN (
...
+						if err := tx.Exec(createSQL).Error; err != nil {
+							return fmt.Errorf("creating new table: %w", err)
+						}
+					}
+
+					// Copy data directly using SQL
+					dataCopySQL := []string{
+						`INSERT INTO users (id, name, display_name, email, provider_identifier, provider, profile_pic_url, created_at, updated_at, deleted_at)
+             SELECT id, name, display_name, email, provider_identifier, provider, profile_pic_url, created_at, updated_at, deleted_at 
+             FROM users_old WHERE EXISTS (SELECT 1 FROM users_old)`,
```
Comment (nblock) [inline hscontrol/db/db.go]:
```suggestion
             FROM users_old`,
```

File: hscontrol/db/db.go [hunk 20]
```diff
@@ -718,13 +733,239 @@ AND auth_key_id NOT IN (
...
+
+						`INSERT INTO pre_auth_keys (id, key, user_id, reusable, ephemeral, used, tags, expiration, created_at)
+             SELECT id, key, user_id, reusable, ephemeral, used, tags, expiration, created_at 
+             FROM pre_auth_keys_old WHERE EXISTS (SELECT 1 FROM pre_auth_keys_old)`,
```
Comment (nblock) [inline hscontrol/db/db.go]:
```suggestion
             FROM pre_auth_keys_old`,
```

File: hscontrol/db/db.go [hunk 21]
```diff
@@ -718,13 +733,239 @@ AND auth_key_id NOT IN (
...
+
+						`INSERT INTO api_keys (id, prefix, hash, expiration, last_seen, created_at)
+             SELECT id, prefix, hash, expiration, last_seen, created_at 
+             FROM api_keys_old WHERE EXISTS (SELECT 1 FROM api_keys_old)`,
```
Comment (nblock) [inline hscontrol/db/db.go]:
```suggestion
             FROM api_keys_old`,
```

File: hscontrol/db/db.go [hunk 22]
```diff
@@ -718,13 +733,239 @@ AND auth_key_id NOT IN (
...
+
+						`INSERT INTO nodes (id, machine_key, node_key, disco_key, endpoints, host_info, ipv4, ipv6, hostname, given_name, user_id, register_method, forced_tags, auth_key_id, last_seen, expiry, approved_routes, created_at, updated_at, deleted_at)
+             SELECT id, machine_key, node_key, disco_key, endpoints, host_info, ipv4, ipv6, hostname, given_name, user_id, register_method, forced_tags, auth_key_id, last_seen, expiry, approved_routes, created_at, updated_at, deleted_at 
+             FROM nodes_old WHERE EXISTS (SELECT 1 FROM nodes_old)`,
```
Comment (nblock) [inline hscontrol/db/db.go]:
```suggestion
             FROM nodes_old`,
```

File: hscontrol/db/db.go [hunk 23]
```diff
@@ -718,13 +733,239 @@ AND auth_key_id NOT IN (
...
+
+						`INSERT INTO policies (id, data, created_at, updated_at, deleted_at)
+             SELECT id, data, created_at, updated_at, deleted_at 
+             FROM policies_old WHERE EXISTS (SELECT 1 FROM policies_old)`,
```
Comment (nblock) [inline hscontrol/db/db.go]:
```suggestion
             FROM policies_old`,
```

**Review by nblock — CHANGES_REQUESTED** [review]

**Review by nblock — APPROVED** [review]
